File: conversion/layers/resource/appearance.py
# @author: Davy Bellens

"""Resource layer: Presentation Appearance and Material schemas."""

def _rgb_of(colour):
    if colour is None or not colour.is_a("IfcColourRgb"):
        return None
    try:
        return (float(colour.Red), float(colour.Green), float(colour.Blue))
    except (TypeError, ValueError, AttributeError):
        return None


def surface_style_rgba(style):
    if style is None or not style.is_a("IfcSurfaceStyle"):
        return None

    for shading in style.Styles or []:
        if not shading.is_a("IfcSurfaceStyleShading"):
            continue

        rgb = _rgb_of(getattr(shading, "SurfaceColour", None))
        if rgb is None:
            continue

        alpha = 1.0
        transparency = getattr(shading, "Transparency", None)
        if transparency is not None:
            try:
                alpha = 1.0 - float(transparency)
            except (TypeError, ValueError):
                alpha = 1.0
        return (rgb[0], rgb[1], rgb[2], max(0.0, min(1.0, alpha)))

    return None


def styled_item_rgba(styled_item):
    if styled_item is None:
        return None

    for entry in styled_item.Styles or []:
        if entry.is_a("IfcPresentationStyleAssignment"):
            for inner in entry.Styles or []:
                rgba = surface_style_rgba(inner)
                if rgba is not None:
                    return rgba
        else:
            rgba = surface_style_rgba(entry)
            if rgba is not None:
                return rgba

    return None


def item_rgba(item):
    try:
        styled = getattr(item, "StyledByItem", None)
    except (AttributeError, RuntimeError):
        return None

    for styled_item in styled or []:
        rgba = styled_item_rgba(styled_item)
        if rgba is not None:
            return rgba
    return None


_MATERIAL_SET_LISTS = {
    "IfcMaterialLayerSet": "MaterialLayers",
    "IfcMaterialList": "Materials",
    "IfcMaterialProfileSet": "MaterialProfiles",
    "IfcMaterialConstituentSet": "MaterialConstituents",
}

_MATERIAL_USAGES = {
    "IfcMaterialLayerSetUsage": "ForLayerSet",
    "IfcMaterialProfileSetUsage": "ForProfileSet",
}

_MATERIAL_HOLDERS = ("IfcMaterialLayer", "IfcMaterialProfile",
                     "IfcMaterialConstituent")


def materials_of(definition, depth=0):
    if definition is None or depth > 6:
        return []

    name = definition.is_a()

    if name == "IfcMaterial":
        return [definition]

    if name in _MATERIAL_USAGES:
        return materials_of(getattr(definition, _MATERIAL_USAGES[name], None),
                            depth + 1)

    if name in _MATERIAL_SET_LISTS:
        out = []
        for child in getattr(definition, _MATERIAL_SET_LISTS[name], None) or []:
            out.extend(materials_of(child, depth + 1))
        return out

    if name in _MATERIAL_HOLDERS:
        return materials_of(getattr(definition, "Material", None), depth + 1)

    return []


UNSTYLED_NAME = "ifc_unstyled"
UNSTYLED_RGBA = (0.78, 0.78, 0.78, 1.0)


class Palette:
    def __init__(self, model, min_alpha=0.0, linear=False):
        self.min_alpha = min_alpha
        self.linear = linear
        self.materials = {}
        self._material_rgba = {}
        self._product_rgba = {}
        self._index(model)

    def _index(self, model):
        try:
            definitions = model.by_type("IfcMaterialDefinitionRepresentation")
        except RuntimeError:
            return

        for definition in definitions:
            material = getattr(definition, "RepresentedMaterial", None)
            if material is None:
                continue
            for representation in definition.Representations or []:
                for item in getattr(representation, "Items", None) or []:
                    if not item.is_a("IfcStyledItem"):
                        continue
                    rgba = styled_item_rgba(item)
                    if rgba is not None:
                        self._material_rgba.setdefault(material.id(), rgba)
                        break

    def product_rgba(self, product):
        key = product.id()
        if key in self._product_rgba:
            return self._product_rgba[key]

        rgba = None
        for association in getattr(product, "HasAssociations", None) or []:
            if not association.is_a("IfcRelAssociatesMaterial"):
                continue
            for material in materials_of(
                    getattr(association, "RelatingMaterial", None)):
                rgba = self._material_rgba.get(material.id())
                if rgba is not None:
                    break
            if rgba is not None:
                break

        self._product_rgba[key] = rgba
        return rgba

    def unstyled(self):
        self.materials.setdefault(UNSTYLED_NAME, UNSTYLED_RGBA)
        return UNSTYLED_NAME

    def register(self, rgba):
        if rgba is None:
            return None
        r, g, b, a = rgba
        if self.min_alpha:
            a = max(a, self.min_alpha)
        clamped = tuple(max(0.0, min(1.0, c)) for c in (r, g, b, a))
        name = "ifc_%02X%02X%02X%02X" % tuple(
            int(round(c * 255)) for c in clamped)
        self.materials.setdefault(name, clamped)
        return name

    def write_mtl(self, path, source=""):
        with open(path, "w") as mtl:
            mtl.write("# materials resolved from %s\n" % source)
            mtl.write("# IfcSurfaceStyleRendering -> Kd, Transparency -> d\n")
            for name in sorted(self.materials):
                r, g, b, a = self.materials[name]
                if self.linear:
                    r, g, b = (srgb_to_linear(c) for c in (r, g, b))
                mtl.write("\nnewmtl %s\n" % name)
                mtl.write("Kd %.6f %.6f %.6f\n" % (r, g, b))
                mtl.write("Ka 0.000000 0.000000 0.000000\n")
                mtl.write("Ks 0.000000 0.000000 0.000000\n")
                mtl.write("Ns 10.0\n")
                mtl.write("d %.4f\n" % a)
                mtl.write("illum %d\n" % (4 if a < 0.999 else 2))
        return len(self.materials)


def srgb_to_linear(c):
    return c / 12.92 if c <= 0.04045 else ((c + 0.055) / 1.055) ** 2.4

File: UnityPackage/com.byvad.ifc-import/Runtime/Layers/Resource/Appearance.cs
using System;
using System.Collections.Generic;
using Conversion.Ifc;

namespace Conversion.Layers.Resource
{
    /// <summary>Resource layer: Presentation Appearance and Material schemas.</summary>
    public sealed class Appearance
    {
        public static readonly Rgba UnstyledColour = new Rgba(0.78, 0.78, 0.78, 1.0);
        public const string UnstyledName = "ifc_unstyled";

        private const int MaxMaterialDepth = 6;

        private static Rgba? RgbOf(IfcEntity colour)
        {
            if (colour == null || !colour.IsA("IfcColourRgb"))
            {
                return null;
            }
            if (!colour.TryDouble("Red", out double r) ||
                !colour.TryDouble("Green", out double g) ||
                !colour.TryDouble("Blue", out double b))
            {
                return null;
            }
            return new Rgba(r, g, b, 1.0);
        }

        public static Rgba? SurfaceStyleRgba(IfcEntity style)
        {
            if (style == null || !style.IsA("IfcSurfaceStyle"))
            {
                return null;
            }

            foreach (IfcEntity shading in style.Entities("Styles"))
            {
                // IfcSurfaceStyleRendering derives from IfcSurfaceStyleShading, so
                // the supertype test catches both.
                if (!shading.IsA("IfcSurfaceStyleShading"))
                {
                    continue;
                }

                Rgba? rgb = RgbOf(shading.Entity("SurfaceColour"));
                if (!rgb.HasValue)
                {
                    continue;
                }

                double alpha = 1.0;
                if (shading.TryDouble("Transparency", out double transparency))
                {
                    alpha = 1.0 - transparency;
                }
                alpha = Math.Max(0.0, Math.Min(1.0, alpha));

                return new Rgba(rgb.Value.R, rgb.Value.G, rgb.Value.B, alpha);
            }
            return null;
        }

        public static Rgba? StyledItemRgba(IfcEntity styledItem)
        {
            if (styledItem == null)
            {
                return null;
            }

            foreach (IfcEntity entry in styledItem.Entities("Styles"))
            {
                if (entry.IsA("IfcPresentationStyleAssignment"))
                {
                    foreach (IfcEntity inner in entry.Entities("Styles"))
                    {
                        Rgba? nested = SurfaceStyleRgba(inner);
                        if (nested.HasValue)
                        {
                            return nested;
                        }
                    }
                }
                else
                {
                    Rgba? direct = SurfaceStyleRgba(entry);
                    if (direct.HasValue)
                    {
                        return direct;
                    }
                }
            }
            return null;
        }

        /// <summary>The style attached directly to a representation item, if any.</summary>
        public Rgba? ItemRgba(IfcEntity item)
        {
            if (item == null)
            {
                return null;
            }
            foreach (IfcEntity styledItem in item.Inverse("StyledByItem"))
            {
                Rgba? rgba = StyledItemRgba(styledItem);
                if (rgba.HasValue)
                {
                    return rgba;
                }
            }
            return null;
        }

        private static readonly Dictionary<string, string> SetLists = new Dictionary<string, string>
        {
            { "IfcMaterialLayerSet", "MaterialLayers" },
            { "IfcMaterialList", "Materials" },
            { "IfcMaterialProfileSet", "MaterialProfiles" },
            { "IfcMaterialConstituentSet", "MaterialConstituents" },
        };

        private static readonly Dictionary<string, string> Usages = new Dictionary<string, string>
        {
            { "IfcMaterialLayerSetUsage", "ForLayerSet" },
            { "IfcMaterialProfileSetUsage", "ForProfileSet" },
        };

        private static readonly HashSet<string> Holders = new HashSet<string>
        {
            "IfcMaterialLayer", "IfcMaterialProfile", "IfcMaterialConstituent",
        };

        /// <summary>Flatten whatever material structure a product points at into plain materials.</summary>
        public static List<IfcEntity> MaterialsOf(IfcEntity definition, int depth = 0)
        {
            var result = new List<IfcEntity>();
            if (definition == null || depth > MaxMaterialDepth)
            {
                return result;
            }

            string name = definition.IsA();

            if (name == "IfcMaterial")
            {
                result.Add(definition);
                return result;
            }

            if (Usages.TryGetValue(name, out string usageAttribute))
            {
                result.AddRange(MaterialsOf(definition.Entity(usageAttribute), depth + 1));
                return result;
            }

            if (SetLists.TryGetValue(name, out string listAttribute))
            {
                foreach (IfcEntity child in definition.Entities(listAttribute))
                {
                    result.AddRange(MaterialsOf(child, depth + 1));
                }
                return result;
            }

            if (Holders.Contains(name))
            {
                result.AddRange(MaterialsOf(definition.Entity("Material"), depth + 1));
            }
            return result;
        }
    }

    /// <summary>
    /// One colour index per model: materials resolved once, products cached on demand.
    /// </summary>
    public sealed class Palette
    {
        private readonly double _minAlpha;
        private readonly bool _linear;
        private readonly Dictionary<int, Rgba> _materialRgba = new Dictionary<int, Rgba>();
        private readonly Dictionary<int, Rgba?> _productRgba = new Dictionary<int, Rgba?>();
        private readonly Dictionary<string, Rgba> _registered =
            new Dictionary<string, Rgba>(StringComparer.Ordinal);

        /// <param name="minAlpha">Floor on opacity. IFC glazing is routinely authored at
        /// Transparency 1.0, which resolves to an invisible surface.</param>
        /// <param name="linear">Convert from sRGB to linear when writing out.</param>
        public Palette(IfcModel model, double minAlpha = 0.0, bool linear = false)
        {
            _minAlpha = minAlpha;
            _linear = linear;
            Index(model);
        }

        public IReadOnlyDictionary<string, Rgba> Materials => _registered;

        private void Index(IfcModel model)
        {
            foreach (IfcEntity definition in model.ByType("IfcMaterialDefinitionRepresentation"))
            {
                IfcEntity material = definition.Entity("RepresentedMaterial");
                if (material == null)
                {
                    continue;
                }
                Rgba? rgba = FirstStyledRgba(definition);
                if (rgba.HasValue && !_materialRgba.ContainsKey(material.Id))
                {
                    _materialRgba[material.Id] = rgba.Value;
                }
            }
        }

        /// <summary>The first styled item's colour found across this definition's representations, in document order.</summary>
        private static Rgba? FirstStyledRgba(IfcEntity definition)
        {
            foreach (IfcEntity representation in definition.Entities("Representations"))
            {
                foreach (IfcEntity item in representation.Entities("Items"))
                {
                    if (!item.IsA("IfcStyledItem"))
                    {
                        continue;
                    }
                    Rgba? rgba = Appearance.StyledItemRgba(item);
                    if (rgba.HasValue)
                    {
                        return rgba;
                    }
                }
            }
            return null;
        }

        /// <summary>The colour a product inherits from its associated material.</summary>
        public Rgba? ProductRgba(IfcEntity product)
        {
            if (_productRgba.TryGetValue(product.Id, out Rgba? cached))
            {
                return cached;
            }

            Rgba? resolved = FindMaterialRgba(product);
            _productRgba[product.Id] = resolved;
            return resolved;
        }

        private Rgba? FindMaterialRgba(IfcEntity product)
        {
            foreach (IfcEntity association in product.Inverse("HasAssociations"))
            {
                if (!association.IsA("IfcRelAssociatesMaterial"))
                {
                    continue;
                }
                foreach (IfcEntity material in Appearance.MaterialsOf(association.Entity("RelatingMaterial")))
                {
                    if (_materialRgba.TryGetValue(material.Id, out Rgba hit))
                    {
                        return hit;
                    }
                }
            }
            return null;
        }

        public string Unstyled()
        {
            if (!_registered.ContainsKey(Appearance.UnstyledName))
            {
                _registered[Appearance.UnstyledName] = Appearance.UnstyledColour;
            }
            return Appearance.UnstyledName;
        }

        /// <summary>Intern a colour and return its stable name, matching the Python's ifc_RRGGBBAA.</summary>
        public string Register(Rgba? rgba)
        {
            if (!rgba.HasValue)
            {
                return null;
            }

            Rgba value = rgba.Value;
            double a = _minAlpha > 0.0 ? Math.Max(value.A, _minAlpha) : value.A;

            double Clamp(double c) => Math.Max(0.0, Math.Min(1.0, c));
            var clamped = new Rgba(Clamp(value.R), Clamp(value.G), Clamp(value.B), Clamp(a));

            string name = clamped.HexName();

            if (!_registered.ContainsKey(name))
            {
                _registered[name] = clamped;
            }
            return name;
        }

        public static double SrgbToLinear(double c) =>
            c <= 0.04045 ? c / 12.92 : Math.Pow((c + 0.055) / 1.055, 2.4);

        /// <summary>
        /// Write a Wavefront .mtl. Not needed for Unity, but it keeps the OBJ path
        /// alive and makes the C# output diffable against the Python's.
        /// </summary>
        public int WriteMtl(string path, string source = "")
        {
            var names = new List<string>(_registered.Keys);
            names.Sort(StringComparer.Ordinal);

            using (var writer = new System.IO.StreamWriter(path))
            {
                writer.Write("# materials resolved from " + source + "\n");
                writer.Write("# IfcSurfaceStyleRendering -> Kd, Transparency -> d\n");
                foreach (string name in names)
                {
                    Rgba c = _registered[name];
                    double r = c.R, g = c.G, b = c.B;
                    if (_linear)
                    {
                        r = SrgbToLinear(r);
                        g = SrgbToLinear(g);
                        b = SrgbToLinear(b);
                    }
                    writer.Write("\nnewmtl " + name + "\n");
                    writer.Write(string.Format(System.Globalization.CultureInfo.InvariantCulture,
                        "Kd {0:F6} {1:F6} {2:F6}\n", r, g, b));
                    writer.Write("Ka 0.000000 0.000000 0.000000\n");
                    writer.Write("Ks 0.000000 0.000000 0.000000\n");
                    writer.Write("Ns 10.0\n");
                    writer.Write(string.Format(System.Globalization.CultureInfo.InvariantCulture,
                        "d {0:F4}\n", c.A));
                    writer.Write("illum " + (c.A < 0.999 ? 4 : 2) + "\n");
                }
            }
            return _registered.Count;
        }
    }
}
